sopel/plugins/capabilities.py

```diff
@@ -0,0 +1,456 @@
+        advertised by the server.
+        """
+        plugin_caps = self._registered.setdefault(request.cap_req, {})
+        plugin_caps[plugin_name] = (
+            request, False,
+        )
+        LOGGER.debug('Capability Request registered: %s', str(request))
+
+    def request_available(
+        self,
+        bot: Sopel,
+        available_capabilities: Iterable[str],
+    ) -> None:
+        """Request available capabilities.
+
+        :param bot: the bot instance used to send capability requests
+        :param available_capabilities: available capabilities
+
+        This sends ``CAP REQ`` commands for requests that can be made, i.e.
+        all the requested capabilities (with or without prefix) must be
+        available for Sopel to send the request.
+
+        Requests made are stored as requested; others are ignored::
+
+            >>> # manager.register => cap1 and cap2, not cap3
+            >>> manager.request_available(bot, ('cap1', 'cap3'))
+            >>> manager.is_requested(('cap1',))
+            True
+            >>> manager.is_requested(('cap2',))
+            False
+            >>> manager.is_requested(('cap3',))
+            False
+
+        .. important::
+
+            The capability request ``('cap1', '-cap2')`` means "enable cap1 and
+            disable cap2", and the request will be acknowledged or denied at
+            once. If the server doesn't advertise any of these capabilities,
+            the client **should not** send a request.
+
+            As a result, Sopel will send a request to enable or disable
+            a capability only if it is advertised first. If a request is
+            never made, its callback will never be called, because there
+            won't be any related ACK/NAK message from the server.
+
+            Plugin authors should not use the request's callback to activate or
+            deactivate features, and instead check the bot's capabilities every
+            time they need to.
+
+            The only exception to that is when the plugin needs to perform an
+            operation while negotiating the capabilities (such as SASL auth).
+
+        """
+        # Requesting capabilities when they are available
+        capabilities = set(available_capabilities)
+        for cap_req, plugin_requests in self._registered.items():
+            # request only if all are available
+            cap_req_enables = {
+                cap.lstrip('-')
+                for cap in cap_req
+            }
+            text = ' '.join(cap_req)
+            if cap_req_enables <= capabilities:
+                self._registered[cap_req] = {
+                    plugin_name: (handler_info[0], False)
+                    for plugin_name, handler_info in plugin_requests.items()
+                }
+                self._requested.add(cap_req)
+                LOGGER.debug('Capability negotiation request: "%s".', text)
+                bot.write(('CAP', 'REQ'), text=text)
```

Comment:
[A discussion above indicates](https://github.com/sopel-irc/sopel/pull/2341/files#r1141195054) that this method should be responsible for ensuring that the resulting `CAP REQ` subcommand isn't too long to be acknowledged in a single line, per a SHOULD in the spec:

> Hey, I found a [recommendation in the spec](https://ircv3.net/specs/extensions/capability-negotiation.html#the-cap-req-subcommand):
> 
> > Clients SHOULD ensure that their list of requested capabilities is not too long to be replied to with a single ACK or NAK message. If a REQ’s final parameter gets sufficiently large (approaching the 510 byte limit), clients SHOULD instead send multiple REQ subcommands.
> 
> So, Sopel is the Client. Sopel SHOULD instead send multiple REQ subcommands.

Reply:
Done!